### PR TITLE
implement close/stop and fix spec timing issue

### LINKED ISSUE #9: Tests are broken
From: http://build-eu-00.elastic.co/view/LS%20Plugins/view/LS%20Inputs/job/logstash-plugin-input-udp-unit/9/

```
Bundle complete! 2 Gemfile dependencies, 34 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :integration=>true, :windows=>true}
..F

Failures:

  1) LogStash::Inputs::Udp receive should receive events been generated
     Failure/Error: client.send("msg #{i}")
     SocketError:
       send: name or service not known
     # ./spec/support/client.rb:28:in `send'
     # ./spec/inputs/udp_spec.rb:32:in `events'
     # ./spec/inputs/udp_spec.rb:31:in `events'
     # ./spec/spec_helper.rb:12:in `input'
     # ./spec/inputs/udp_spec.rb:30:in `events'
     # ./spec/inputs/udp_spec.rb:42:in `(root)'

Finished in 0.631 seconds (files took 1.47 seconds to load)
3 examples, 1 failure
```

### LINKED ISSUE #11: shutdown refactor issues
there are issues that need to be fixed related to the shutdown refactor. 

- a `stop` method should be implemented that force-closes the socket to make sure it will not ne stuck in IO somewhere regardless if we are using select... for safety.

- I would change the current close method to 
```ruby
def close
  @udp.close rescue nil
end 
```
and add an identical stop method
```ruby
def stop
  @udp.close rescue nil
end 
```

note that with a proper `stop` the current `close` method is useless, unless we move the UDPSocket creation in the register method as proposed in #10 where we could have a situation where the plugin `register` method is called but not the `run` & `stop` and then only `close`, like in specs in which case both `close` and `stop` are required.

- there is an error in the `spec_helper.rb` at https://github.com/logstash-plugins/logstash-input-udp/blob/12050767da47a63cf03b399847b8d38e7578765c/spec/spec_helper.rb#L17

`plugin.stop` should be rewritten as

```ruby
plugin.do_stop
input_thread.join
```

otherwise the `stop?` method will never return `true` and there could be race conditions with the input thread.

### PR DESCRIPTION
**Pull request by jsvd**
fixes #11
fixes #9

### CODE REVIEW

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -2,6 +2,10 @@
 require "logstash/devutils/rspec/spec_helper"
 require 'logstash/inputs/udp'
 
+class LogStash::Inputs::Udp
+  attr_reader :udp
+end
```
Comment (colinsurprenant) [inline spec/spec_helper.rb]: I am sure there is a valid reason for adding this here :P  - could you enlighten me and document it?

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -9,12 +13,15 @@ def input(plugin, size, &block)
     input_thread = Thread.new do
       plugin.run(queue)
     end
+    # because the udp socket is created and bound during #run
+    # we must ensure that it is open before sending data
+    sleep 0.1 until (plugin.udp && !plugin.udp.closed?)
```
Comment (colinsurprenant) [inline spec/spec_helper.rb]: oh ^^ I see. but isn't this going to change with #10 ?

**Comment by colinsurprenant** [comment]
I am a bit confused. was this PR meant to be a first step replacing #10 for now ?

**Comment by colinsurprenant** [comment]
ok, re-read your comment in #11 and I understand this is a minimal refactor for shutdown to avoid the yak shave introduced by #10.

File: spec/spec_helper.rb (hunk 2 above)
Comment (jsvd) [inline spec/spec_helper.rb]: it will. eventually. until then this needs to be fixed.

**Comment by colinsurprenant** [comment]
LGTM

**Pull request closed without merging**